plugins/mac-capture/mac-window-capture.m

```diff
@@ -13,71 +13,91 @@
 
 	struct cocoa_window window;
 
-	//CGRect              bounds;
-	//CGWindowListOption  window_option;
 	CGWindowImageOption image_option;
 
-	CGColorSpaceRef color_space;
-
 	DARRAY(uint8_t) buffer;
 
+	uint64_t last_tick;
+	struct obs_source_frame frame;
+
 	pthread_t capture_thread;
 	os_event_t *capture_event;
 	os_event_t *stop_event;
+
+	pthread_mutex_t mutex;
 };
 
 static CGImageRef get_image(struct window_capture *wc)
 {
-	NSArray *arr = (NSArray *)CGWindowListCreate(
-		kCGWindowListOptionIncludingWindow, wc->window.window_id);
-	[arr autorelease];
+	CFMutableArrayRef arr =
+		CFArrayCreateMutable(kCFAllocatorDefault, 0, NULL);
+	CFArrayAppendValue(arr, (void *)(uintptr_t)wc->window.window_id);
+
+	CGImageRef image = CGWindowListCreateImageFromArray(
+		CGRectNull, (CFArrayRef)arr, wc->image_option);
 
-	if (!arr.count && !find_window(&wc->window, NULL, false))
-		return NULL;
+	if (!image && find_window(&wc->window, NULL))
+		image = get_image(wc);
 
-	return CGWindowListCreateImage(CGRectNull,
-				       kCGWindowListOptionIncludingWindow,
-				       wc->window.window_id, wc->image_option);
+	CFRelease(arr);
+
+	return image;
 }
 
 static inline void capture_frame(struct window_capture *wc)
 {
 	uint64_t ts = os_gettime_ns();
-	CGImageRef img = get_image(wc);
-	if (!img)
+	CGImageRef image = get_image(wc);
+
+	if (!image)
+		return;
+
+	size_t width = CGImageGetWidth(image);
+	size_t height = CGImageGetHeight(image);
+	size_t bytes_per_row = CGImageGetBytesPerRow(image);
+
+	if ((!width && !height) || CGImageGetBitsPerPixel(image) != 32 ||
+	    CGImageGetBitsPerComponent(image) != 8) {
+		CGImageRelease(image);
 		return;
+	}
+
+	// float fps = 1e9 / (ts - wc->frame.timestamp);
+	// NSLog(@"FPS %.4f", fps);
+
+	CGDataProviderRef provider = CGImageGetDataProvider(image);
+	CFDataRef data = CGDataProviderCopyData(provider);
 
-	size_t width = CGImageGetWidth(img);
-	size_t height = CGImageGetHeight(img);
+	da_resize(wc->buffer, CFDataGetLength(data));
+	uint8_t *buffer = wc->buffer.array;
 
-	CGRect rect = {{0, 0}, {width, height}};
-	da_resize(wc->buffer, width * height * 4);
-	uint8_t *data = wc->buffer.array;
+	CFDataGetBytes(data, CFRangeMake(0, CFDataGetLength(data)), buffer);
```

Comment:
Yes, thank you!  I tried looking for `CFDataGetBytePtr()` and just couldn't find it at the time, because I wanted to eliminate what should have been an unnecessary copy of the data, but hadn't found it at the time.  This will definitely help simplify things.